### ISSUE TITLE
Unable to resolve react-native location in node_modules.

### ISSUE THREAD

**Issue by dhatGuy** [software-mansion/react-native-reanimated#3836]
### Description

I tried building my app but kept seeing this error
![image](https://user-images.githubusercontent.com/51405947/205683237-509458b2-7291-49b5-9a13-f8a9c8de1148.png)

This issue was said to be addressed in [3.0.0-rc.4](https://github.com/software-mansion/react-native-reanimated/releases/tag/3.0.0-rc.4) still persists for me

### Steps to reproduce

1. upgrade to expo 47
2. expo run:android

### Snack or a link to a repository

N/A

### Reanimated version

3.0.0-rc.8

### React Native version

0.70.5

### Platforms

Android

### JavaScript runtime

Hermes

### Workflow

Expo bare workflow

### Architecture

None

### Build type

None

### Device

Android emulator

### Acknowledgements

Yes

**Comment by Calvin-Octoco** [comment]
Getting the same issue with Android when using Expo EAS build:
```
[stderr] FAILURE: Build failed with an exception.
[stderr] * Where:
[stderr] Build file '/home/expo/workingdir/build/apps/paga-app/node_modules/react-native-reanimated/android/build.gradle' line: 114
[stderr] * What went wrong:
[stderr] A problem occurred evaluating project ':react-native-reanimated'.
[stderr] > [Reanimated] Unable to resolve react-native location in node_modules. You should project extension property (in app/build.gradle) `REACT_NATIVE_NODE_MODULES_DIR` with path to react-native.
```

**Comment by qroac** [comment]
Out of a sudden I get the very same error on Expo 45.
I tried to re-build a git version that worked on Nov. 21. Now it fails with this error.
Only difference I could spot in the build logs:
- on Nov. 21 expo worker had node v16.13.2 as default
- now it has node v16.18.1 as default and installs my required v16.13.2 later

Expo forum post: https://forums.expo.dev/t/gradlew-job-fails-to-resolve-react-native-monorepo-build-already-worked-on-nov-21/68274

**Comment by a-eid** [comment]
I am getting the same error on the latest version of reanimated as well as the RC one.

**Comment by chriscoomber** [comment]
Does anyone know a workaround to this when using managed expo? I would rather stick to reanimated 2.12.x since 3.x isn't officially supported with expo yet.

I'm using a monorepo setup, so my react-native is in `project/node_modules/react-native` instead of  `project/apps/app-name/node_modules/react-native` Expo go works fine (e.g. `expo start`) but eas builds (`eas build --profile preview -p android`) fail with the error from this issue. I'm happy to provide the environment variable `REACT_NATIVE_NODE_MODULES_DIR` but providing it in eas.json doesn't seem to do anything.

**Comment by dhatGuy** [comment]
> Does anyone know a workaround to this when using managed expo? I would rather stick to reanimated 2.12.x since 3.x isn't officially supported with expo yet.
> 
> I'm using a monorepo setup, so my react-native is in `monorepo/node_modules/react-native` instead of `monorepo/apps/my-app/node_modules/react-native` Expo go works fine (e.g. `expo start`) but eas builds (`eas build --profile preview -p android`) fail with the error from this issue. I'm happy to provide the environment variable `REACT_NATIVE_NODE_MODULES_DIR` but providing it in eas.json doesn't seem to do anything.
> 
> Edit: I have worked around this for my use case. My workaround is to _not_ hoist `react-native` for this app. In `monorepo/apps/my-app/package.json` I added:
> 
> ```json
>   "workspaces": {
>     "nohoist": ["react-native", "react-native/**"]
>   }
> ```

Okay, this workaround works with eas but doesn't work when building locally i.e `expo run:android`

**Comment by chriscoomber** [comment]
@dhatGuy Yes, you're right unfortunately. My current situation is that I add those nohoist rules to my `package.json` just before doing an EAS build, and remove them after - I don't check them into version control. It's not ideal.

For some reason, for me `expo start` (or `expo run:android`) works fine for me without them, and breaks with them, whereas `eas build` is the opposite. So, I just add them for `eas build`. I'm afraid I don't have any advice for building locally (e.g. `expo run:android`), as it works fine for me.

**Comment by SGeri** [comment]
Hi. I had the sam issue while using `react-native-gesture-handler` in a monorepo setup. I couldn't figure it out why this package was installed always in the app's node_modules instead of the root's one. I tried `npm i `and `npm ci` several times but the `react-native-gesture-handler` package was not behaving like I wanted it to. 

I managed to solve this issue by deleting the `package-lock.json` and reinstalling the npm packages. This way, the `react-native-gesture-handler` was in the correct `node_modules` and everything worked as a charm.
I hope this works for others too!

Also, you guys mentioned that using the `"nohoist"` option solves this problem for most. That feature is only available in yarn, unfortunately not with npm workspaces.

**Comment by piaskowyk** [comment]
Hey! For you project configuration, to solve this issue, you should project extension property (in app/build.gradle) `REACT_NATIVE_NODE_MODULES_DIR` with path to react-native.

**Comment by piaskowyk** [comment]
I hope that it solved your issue, let me know if you still have a problem with it.

**Comment by stretch0** [comment]
@piaskowyk is `REACT_NATIVE_NODE_MODULES_DIR` just an environment variable that's supposed to be exported?

I am doing the following but still get the same error `Unable to resolve react-native location in node_modules...`

```
export REACT_NATIVE_NODE_MODULES_DIR=../node_modules
```

My monorepo structure is fairly straight forward

```
├ node_modules // react-native seems to get installed in here for some reason so I guess I need to set `REACT_NATIVE_NODE_MODULES_DIR=../node_modules`
├ package.json
├ my-app
     ├ node_modules
     ├ package.json
```

**Comment by chrisdrifte** [comment]
I ran into this issue in my monorepo, it seems the following code block in build.grade is relevant:

```
def resolveReactNativeDirectory() {
    def reactNativeLocation = safeAppExtGet("REACT_NATIVE_NODE_MODULES_DIR", null)
    if (reactNativeLocation != null) {
        return file(reactNativeLocation)
    }

    if (isDeveloperMode() || isPackageBuild()) {
        return file("$projectDir/../${getPlaygroundAppName()}/node_modules/react-native")
    }

    // monorepo workaround
    // react-native can be hoisted or in project's own node_modules
    def reactNativeFromProjectNodeModules = file("${rootProject.projectDir}/../node_modules/react-native")
    if (reactNativeFromProjectNodeModules.exists()) {
        return reactNativeFromProjectNodeModules
    }
    
    def reactNativeFromNodeModulesWithReanimated = file("${projectDir}/../../react-native")
    if (reactNativeFromNodeModulesWithReanimated.exists()) {
        return reactNativeFromNodeModulesWithReanimated
    }

    throw new GradleException(
        "[Reanimated] Unable to resolve react-native location in " +
        "node_modules. You should project extension property (in app/build.gradle) " +
        "`REACT_NATIVE_NODE_MODULES_DIR` with path to react-native."
    )
}
```

I had `react-native-reanimated` in nohoist in my root `package.json`, as a workaround to a different build issue (for iOS). 

❌ Adding `react-native` to nohoist didn't solve my issue
✅ Removing `react-native-reanimated` from nohoist did.

**Comment by enisze** [comment]
Hey I got the same issue now:
<img width="580" alt="image" src="https://github.com/software-mansion/react-native-reanimated/assets/15271433/e8a20c04-e40b-4021-aacd-f867bcf6a1f7">

Using expo 49 and react-native-reanimated 3.6.1

**Comment by pitops** [comment]
For me in a monorepo with apps/mobile folder for expo I had to add to build.gradle the code below above the react {}

```
project.ext {
    REACT_NATIVE_NODE_MODULES_DIR = file("../../../../node_modules/react-native")
}
```

I am using bun for bundler.
